File: server/services/queue.py
"""Message queue abstraction and Redis Streams implementation.

Provides a backend-agnostic ``MessageQueue`` ABC so the bridge and SSE
layers never touch queue internals directly.  The default (and currently
only) implementation is ``RedisStreamQueue`` which wraps Redis Streams
commands (XADD / XREADGROUP / XACK / XTRIM).

A ``create_queue()`` factory selects the concrete implementation based
on configuration so callers only depend on the abstract interface.
"""

from abc import ABC, abstractmethod
from typing import Optional

from redis.asyncio import Redis
from redis.asyncio.cluster import RedisCluster

from infra.log import logger


# ---------------------------------------------------------------------------
# Abstract base class
# ---------------------------------------------------------------------------

class MessageQueue(ABC):
    """Backend-agnostic message queue interface."""

    @abstractmethod
    async def publish(self, queue_name: str, message: str) -> str:
        """Append *message* to *queue_name*.

        Returns:
            A backend-specific message ID (e.g. a Redis Stream entry ID).
        """
        ...

    @abstractmethod
    async def consume(
        self,
        queue_name: str,
        group: str,
        consumer: str,
        block_ms: int = 5000,
    ) -> Optional[tuple[str, str]]:
        """Block-read one message from *queue_name*.

        Returns:
            ``(message_id, payload)`` or ``None`` when *block_ms* elapses
            with no new messages.
        """
        ...

    @abstractmethod
    async def ack(self, queue_name: str, group: str, message_id: str) -> None:
        """Acknowledge successful processing of *message_id*."""
        ...

    @abstractmethod
    async def delete_message(self, queue_name: str, message_id: str) -> None:
        """Remove *message_id* from the stream to reclaim memory."""
        ...

    @abstractmethod
    async def delete_queue(self, queue_name: str) -> None:
        """Delete the queue and all its data."""
        ...

    @abstractmethod
    async def purge(self, queue_name: str) -> None:
        """Remove all messages but keep the queue structure."""
        ...

    @abstractmethod
    async def ensure_group(self, queue_name: str, group: str) -> None:
        """Create the consumer group if it does not already exist.

        New groups start at offset ``$`` (only future messages).
        """
        ...


# ---------------------------------------------------------------------------
# Redis Streams implementation
# ---------------------------------------------------------------------------

class RedisStreamQueue(MessageQueue):
    """``MessageQueue`` backed by Redis Streams.

    Each *queue_name* maps to a single Redis Stream key.  Consumer groups
    are created lazily via :meth:`ensure_group` which auto-creates the
    underlying key if absent.

    Compatible with both Redis standalone and Redis Cluster (every
    operation touches a single key).
    """

    def __init__(self, redis: Redis, *, max_stream_len: int = 1000) -> None:
        self._redis = redis
        self._max_len = max_stream_len

    # -- publish -------------------------------------------------------------

    async def publish(self, queue_name: str, message: str) -> str:
        entry_id: str = await self._redis.xadd(
            queue_name,
            {"data": message},
            maxlen=self._max_len,
            approximate=True,
        )
        logger.debug("Queue publish: stream={} msg_id={}", queue_name, entry_id)
        return entry_id

    # -- consume -------------------------------------------------------------

    async def consume(
        self,
        queue_name: str,
        group: str,
        consumer: str,
        block_ms: int = 5000,
    ) -> Optional[tuple[str, str]]:
        try:
            result = await self._redis.xreadgroup(
                groupname=group,
                consumername=consumer,
                streams={queue_name: ">"},
                count=1,
                block=block_ms,
            )
        except Exception as exc:
            # NOGROUP — group hasn't been created yet (race on first call)
            err_msg = str(exc)
            if "NOGROUP" in err_msg:
                logger.warning("Queue NOGROUP: stream={} group={}, recreated", queue_name, group)
                await self.ensure_group(queue_name, group)
                return None
            raise

        if not result:
            return None

        # result structure: [[stream_name, [(entry_id, {field: value})]]]
        _stream_name, entries = result[0]
        if not entries:
            return None

        entry_id, fields = entries[0]
        logger.debug("Queue consume: stream={} group={} msg_id={}", queue_name, group, entry_id)
        return (entry_id, fields.get("data", ""))

    # -- ack -----------------------------------------------------------------

    async def ack(self, queue_name: str, group: str, message_id: str) -> None:
        await self._redis.xack(queue_name, group, message_id)
        logger.debug("Queue ack: stream={} msg_id={}", queue_name, message_id)

    # -- delete_message ------------------------------------------------------

    async def delete_message(self, queue_name: str, message_id: str) -> None:
        await self._redis.xdel(queue_name, message_id)
        logger.debug("Queue delete_message: stream={} msg_id={}", queue_name, message_id)

    # -- delete_queue --------------------------------------------------------

    async def delete_queue(self, queue_name: str) -> None:
        await self._redis.delete(queue_name)
        logger.debug("Queue delete_queue: stream={}", queue_name)

    # -- purge ---------------------------------------------------------------

    async def purge(self, queue_name: str) -> None:
        await self._redis.xtrim(queue_name, maxlen=0)
        logger.debug("Queue purge: stream={}", queue_name)

    # -- ensure_group --------------------------------------------------------

    async def ensure_group(self, queue_name: str, group: str) -> None:
        try:
            # Manually ensure the stream key exists before creating the
            # consumer group.  This avoids passing ``mkstream=True`` which
            # can cause encoding issues (bytes vs str) on RedisCluster
            # with ``decode_responses=True``.
            if not await self._redis.exists(queue_name):
                entry_id = await self._redis.xadd(queue_name, {"_init": "1"})
                await self._redis.xdel(queue_name, entry_id)

            # RedisCluster._determine_slot cannot resolve the key
            # position for XGROUP subcommands (neither the high-level
            # xgroup_create() nor execute_command with split tokens work).
            # Bypass slot detection entirely by computing the target node
            # from the key ourselves.
            if isinstance(self._redis, RedisCluster):
                node = self._redis.nodes_manager.get_node_from_slot(
                    self._redis.keyslot(queue_name),
                )
                await self._redis.execute_command(
                    "XGROUP", "CREATE", queue_name, group, "$",
                    target_nodes=node,
                )
            else:
                await self._redis.xgroup_create(
                    queue_name, group, id="$",
                )
        except Exception as exc:
            # BUSYGROUP — group already exists, safe to ignore
            if "BUSYGROUP" in str(exc):
                logger.debug("Queue ensure_group: stream={} group={} (already exists)", queue_name, group)
                return
            raise


# ---------------------------------------------------------------------------
# Factory
# ---------------------------------------------------------------------------

def create_queue(queue_type: str, redis: Redis, **kwargs) -> MessageQueue:
    """Instantiate a :class:`MessageQueue` implementation.

    Args:
        queue_type: ``"redis_stream"`` (only supported value today).
        redis: An initialised async Redis client.
        **kwargs: Forwarded to the concrete constructor
                  (e.g. ``max_stream_len``).
    """
    if queue_type == "redis_stream":
        return RedisStreamQueue(redis, **kwargs)
    raise ValueError(f"Unsupported queue type: {queue_type!r}")

File: backend/internal/service/queue.go
package service

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/redis/go-redis/v9"
	"github.com/rs/zerolog/log"
)

// MessageQueue is the backend-agnostic message queue interface.
type MessageQueue interface {
	Publish(ctx context.Context, queueName, message string) (string, error)
	Consume(ctx context.Context, queueName, group, consumer string, blockMs int) (*QueueMessage, error)
	Ack(ctx context.Context, queueName, group, messageID string) error
	DeleteMessage(ctx context.Context, queueName, messageID string) error
	DeleteQueue(ctx context.Context, queueName string) error
	Purge(ctx context.Context, queueName string) error
	EnsureGroup(ctx context.Context, queueName, group string) error
}

// QueueMessage represents a consumed message.
type QueueMessage struct {
	ID   string
	Data string
}

// RedisStreamQueue implements MessageQueue backed by Redis Streams.
type RedisStreamQueue struct {
	rdb    redis.UniversalClient
	maxLen int64
}

// NewRedisStreamQueue creates a new RedisStreamQueue.
func NewRedisStreamQueue(rdb redis.UniversalClient, maxStreamLen int) *RedisStreamQueue {
	return &RedisStreamQueue{
		rdb:    rdb,
		maxLen: int64(maxStreamLen),
	}
}

func (q *RedisStreamQueue) Publish(ctx context.Context, queueName, message string) (string, error) {
	entryID, err := q.rdb.XAdd(ctx, &redis.XAddArgs{
		Stream: queueName,
		MaxLen: q.maxLen,
		Approx: true,
		Values: map[string]interface{}{"data": message},
	}).Result()
	if err != nil {
		return "", fmt.Errorf("xadd: %w", err)
	}
	log.Debug().Str("stream", queueName).Str("msg_id", entryID).Msg("Queue publish")
	return entryID, nil
}

func (q *RedisStreamQueue) Consume(ctx context.Context, queueName, group, consumer string, blockMs int) (*QueueMessage, error) {
	result, err := q.rdb.XReadGroup(ctx, &redis.XReadGroupArgs{
		Group:    group,
		Consumer: consumer,
		Streams:  []string{queueName, ">"},
		Count:    1,
		Block:    time.Duration(blockMs) * time.Millisecond,
	}).Result()
	if err != nil {
		if err == redis.Nil {
			return nil, nil
		}
		// NOGROUP — group hasn't been created yet
		errMsg := err.Error()
		if strings.Contains(errMsg, "NOGROUP") {
			log.Warn().Str("stream", queueName).Str("group", group).Msg("Queue NOGROUP, recreating")
			_ = q.EnsureGroup(ctx, queueName, group)
			return nil, nil
		}
		return nil, fmt.Errorf("xreadgroup: %w", err)
	}

	if len(result) == 0 || len(result[0].Messages) == 0 {
		return nil, nil
	}

	msg := result[0].Messages[0]
	data, _ := msg.Values["data"].(string)
	log.Debug().Str("stream", queueName).Str("group", group).Str("msg_id", msg.ID).Msg("Queue consume")
	return &QueueMessage{ID: msg.ID, Data: data}, nil
}

func (q *RedisStreamQueue) Ack(ctx context.Context, queueName, group, messageID string) error {
	_, err := q.rdb.XAck(ctx, queueName, group, messageID).Result()
	if err != nil {
		return fmt.Errorf("xack: %w", err)
	}
	log.Debug().Str("stream", queueName).Str("msg_id", messageID).Msg("Queue ack")
	return nil
}

func (q *RedisStreamQueue) DeleteMessage(ctx context.Context, queueName, messageID string) error {
	_, err := q.rdb.XDel(ctx, queueName, messageID).Result()
	if err != nil {
		return fmt.Errorf("xdel: %w", err)
	}
	log.Debug().Str("stream", queueName).Str("msg_id", messageID).Msg("Queue delete_message")
	return nil
}

func (q *RedisStreamQueue) DeleteQueue(ctx context.Context, queueName string) error {
	_, err := q.rdb.Del(ctx, queueName).Result()
	if err != nil {
		return fmt.Errorf("del: %w", err)
	}
	log.Debug().Str("stream", queueName).Msg("Queue delete_queue")
	return nil
}

func (q *RedisStreamQueue) Purge(ctx context.Context, queueName string) error {
	_, err := q.rdb.XTrimMaxLen(ctx, queueName, 0).Result()
	if err != nil {
		return fmt.Errorf("xtrim: %w", err)
	}
	log.Debug().Str("stream", queueName).Msg("Queue purge")
	return nil
}

func (q *RedisStreamQueue) EnsureGroup(ctx context.Context, queueName, group string) error {
	err := q.rdb.XGroupCreateMkStream(ctx, queueName, group, "$").Err()
	if err != nil {
		errMsg := err.Error()
		if strings.Contains(errMsg, "BUSYGROUP") {
			log.Debug().Str("stream", queueName).Str("group", group).Msg("Queue ensure_group: already exists")
			return nil
		}
		return fmt.Errorf("xgroup create: %w", err)
	}
	return nil
}

// NewQueue creates a MessageQueue implementation based on queue type.
func NewQueue(queueType string, rdb redis.UniversalClient, maxStreamLen int) (MessageQueue, error) {
	switch queueType {
	case "redis_stream":
		return NewRedisStreamQueue(rdb, maxStreamLen), nil
	default:
		return nil, fmt.Errorf("unsupported queue type: %q", queueType)
	}
}
